core/store/src/trie/shard_tries.rs

```diff
@@ -156,6 +235,50 @@ impl ShardTries {
         self.get_trie_for_shard_internal(shard_uid, state_root, false, None)
     }
 
+    pub fn get_trie_with_block_hash_for_shard_from_snapshot(
+        &self,
+        shard_uid: ShardUId,
+        state_root: StateRoot,
+        block_hash: &CryptoHash,
+    ) -> Result<Trie, StorageError> {
+        let (store, flat_storage_manager) = {
+            // Taking this lock can last up to 10 seconds, if the snapshot happens to be re-created.
+            match self.0.state_snapshot.try_read() {
+                Ok(guard) => {
+                    if let Some(data) = guard.as_ref() {
+                        if &data.prev_block_hash != block_hash {
+                            return Err(StorageInconsistentState(format!(
+                                "Wrong state snapshot. Requested: {:?}, Available: {:?}",
+                                block_hash, data.prev_block_hash
+                            )));
+                        }
+                        (data.store.clone(), data.flat_storage_manager.clone())
+                    } else {
+                        return Err(StorageInconsistentState(
+                            "No state snapshot available".to_string(),
+                        ));
+                    }
+                }
+                Err(TryLockError::WouldBlock) => {
+                    return Err(StorageInconsistentState(
+                        "Accessing state snapshot would block. Retry in a few seconds.".to_string(),
+                    ));
+                }
+                Err(err) => {
+                    return Err(StorageInconsistentState(format!(
+                        "Can't access state snapshot: {err:?}"
+                    )));
+                }
+            }
+        };
+
+        let cache = TrieCache::new(&self.0.trie_config, shard_uid, true);
```

Comment:
It is assumed that `ShardTries` already has all the caches and creating another one is not a good idea. In the current setup it would duplicate 1 GB cache for shard 3.

Using `caches` would block the main thread, so let's use cache from `view_caches` - it is perfect case for them!